ipldutil/traverser.go

```diff
@@ -233,55 +243,52 @@ func (t *traverser) Shutdown(ctx context.Context) {
 	}
 }
 
-// IsComplete returns true if a traversal is complete
 func (t *traverser) IsComplete() (bool, error) {
-	t.checkState()
+	// If the state is currently held due to an ongoing block load,
+	// block until it's finished or until the traverser stops,
+	// which then enables us to read the fields directly.
+	t.stateMu.Lock()
+	defer t.stateMu.Unlock()
 	return t.isDone, t.completionErr
 }
 
-// CurrentRequest returns the current block load waiting to be fulfilled in order
-// to advance further
 func (t *traverser) CurrentRequest() (ipld.Link, ipld.LinkContext) {
-	t.checkState()
+	// Just like IsComplete.
+	t.stateMu.Lock()
+	defer t.stateMu.Unlock()
 	return t.currentLink, t.currentContext
 }
 
-// Advance advances the traversal with an io.Reader for the next requested block
 func (t *traverser) Advance(reader io.Reader) error {
-	isComplete, _ := t.IsComplete()
-	if isComplete {
-		return errors.New("cannot advance when done")
-	}
+	// Just like IsComplete, block until we're ready to load another block.
+	// We leave it to the other goroutine to unlock the mutex,
+	// once the next StorageReadOpener call comes in or the traversal is done.
+	t.stateMu.Lock()
 
-	select {
-	case <-t.ctx.Done():
-		return ContextCancelError{}
-	case t.awaitRequest <- struct{}{}:
+	if t.isDone {
+		return errors.New("cannot advance when done")
 	}
 
 	select {
 	case <-t.ctx.Done():
 		return ContextCancelError{}
```

Comment:
Same issue here -- if the context is done, what will cause the lock to ever unlock again?